WordPress/gutenberg
# [WIP] Layout: Explore adding in sticky position support within the layout support

## What?


🚧 🚧 🚧 🚧 🚧 WIP: This is nowhere near ready for testing yet 🚧 🚧 🚧 🚧 🚧 

This PR is an early exploration into storing `position` and related data as part of the Layout object, in an effort to come up with a solution for #30121. This expands on ideas earlier explored in #38039.

Related to: #30121

## Why?


It'll be useful for many kinds of sites to be able to set one or two group blocks or template areas to be `sticky` positioned in order to anchor a header or footer to the top or the bottom of the page.

## How?


TBC

Note: much of this is hacked together for now — all the changes in 6.1 files should be moved elsewhere prior to stabilising the PR.

## To-do

* [x] Wire up the `attributes.layout.position` value to style output in both the editor and on the site frontend (e.g. in the layout output in `layout.php`)
* [x] Switch `layout.position` to be an object, as there are likely related values that need to be stored. For example, we need the position type (e.g. `sticky`) as well as the particular edge it should be close to `top`, `bottom`, etc. Also, we might want an offset value if elements are to be set slightly off from being adjacent to the edge of the viewport.
* [x] Include related rules to offset based on the presence of the logged in admin area.
* [ ] In the output include an inferred `z-index` value to ensure that the sticky positioned element sits where it should. Edit: at the moment this is hard-coded.
* [ ] In the editor, figure out how we deal with `z-index` and ensuring that the element can be reached appropriately via List View, clicking on the element, etc, and all inspector / toolbar controls are still behaving as expected
* [ ] Even with a hard-coded z-index value, the block toolbar appears to scroll off the screen when using sticky position
* [ ] Might not be required for an MVP, but me might also want to look into overriding `scroll-padding-top` when setting a header in the site editor (e.g. [this rule](https://github.com/WordPress/wordpress-develop/blob/29b78bd67956f0ef0d7399b81427671164bc5312/src/wp-content/themes/twentysixteen/style.css#L2779))
* [ ] Explore the right place to store the `position` object in block attributes. Should it live in the `layout` object, or in the `style` object? There are trade-offs either way.

## Testing Instructions





***For testing purposes, for now, let's focus on looking just at the Group block***

1. Open up the block editor, add a group block and put a paragraph inside it, and give the group block a background color
2. In the inspector controls, set the Area to top/header, and position to Sticky
3. Scroll down, and the block should stick to the top of the screen
4. Publish the post or page, and on the site frontend, the block should stick to the top as you scroll, and factor in the height of the logged-in admin bar when it is visible

## Screenshots or screencast 

WIPs:

| Editor | Site frontend|
| --- | --- |
| <img src="https://user-images.githubusercontent.com/14988353/195264634-141357c0-e482-43cf-94f2-3d4901c73561.gif" /> | <img src="https://user-images.githubusercontent.com/14988353/195512494-dc6d1b22-e48a-4b6a-99f7-c012bb2fc0f2.gif" /> |
